mlbench_core/controlflow/tensorflow/train_and_eval.py

```diff
@@ -0,0 +1,164 @@
+                          *[m.avg for m in metrics_meter], tracker.best_epoch,
+                          tracker.best_epoch_value))
+
+        # Record training loss and metrics.
+        tracker.records['train_loss'].append(loss_meter.avg)
+        for metric, meter in zip(self.metrics, metrics_meter):
+            tracker.records['train_' + metric['name']].append(meter.avg)
+
+    def valid_one_epoch(self, tracker):
+        self.sess.run(self.data_loader.val_data_init_op)
+
+        num_batches = self.data_loader.num_batches_per_epoch_for_eval
+
+        loss_meter = AverageMeter()
+        metrics_meter = [AverageMeter() for _ in self.metrics]
+
+        for i_batch in range(num_batches):
+            # Get dataset and target from a batch
+            data, target = self.sess.run(self.data_loader.val_next_batch)
+
+            out = self.sess.run(
+                {
+                    "metrics": [m['value'] for m in self.metrics],
+                    "loss": self.loss,
+                },
+                feed_dict={
+                    self.data_loader.inputs: data,
+                    self.data_loader.labels: target,
+                    self.is_training: False
+                })
+
+            # Update tracker
+            loss_meter.update(out["loss"], n=target.shape[0])
+            for meter, o in zip(metrics_meter, out['metrics']):
+                meter.update(o, n=target.shape[0])
+
+            print(("{}/{} Validation loss={:10.3e} | metrics: [" + " ".join
+                   (["{: 10.2f}" for _ in metrics_meter]) + "]")
+                  .format(tracker.current_epoch, i_batch, loss_meter.avg,
+                          *[m.avg for m in metrics_meter]))
+
+        # Record
+        tracker.records['val_loss'].append(loss_meter.avg)
+        for i, metric, meter in zip(range(len(self.metrics)), self.metrics, metrics_meter):
+            metric_name = 'val_' + metric['name']
+
+            # Here we implicitly assume the larger metrics value means better results
+            if i == 0:
+                # primary metrics
+                if (len(tracker.records[metric_name]) == 0 or
+                        meter.avg > max(tracker.records[metric_name])):
+                    tracker.best_epoch = tracker.current_epoch
+                    tracker.best_epoch_value = meter.avg
+
+            tracker.records[metric_name].append(meter.avg)
+
+    def train_and_eval(self, initial_epoch=0, lr_scheduler=None):
+        """Train and evaluate one epoch.
+
+        Args:
+            initial_epoch (int, optional): Defaults to 0. Initial epoch of training.
+            lr_scheduler ([type], optional): Defaults to None. [description]
+
+        Returns:
+            {Namespace}: 
+        """
+
+        # Initialize Variables
+        self.sess.run(tf.group(tf.global_variables_initializer()))
+
+        # Tracker for stats
+        tracker = Tracker()
+        tracker.current_epoch = 0
```

Comment:
I switched the pytorch implementation over to start at epoch 1, since that's what I've seen more often in the wild. We should do the same here.